### PR TITLE
Changed files support for GerritChangeSource and GerritEventLogPoller

### LINKED ISSUE #3996: No "Changed files" in Console view while using GerritChangeSource
We have 2 instances of Buildbot. First instance gets changes from Github (using GitPoller) and the second gets it from Gerrit (using GerritChangeSource).
In case of 1st instance, which uses Github, we have filled "changed files" section. Example:
image
It is ok.

In case of 2nd instance, which uses Gerrit, "changed files" section is empty:
image

We use GerritChangeSource like this:
```
from buildbot.plugins import changes

c['change_source'] = changes.GerritChangeSource(
    "gerrit.example.com",
    "gerrit_user",
    handled_events=["patchset-created", "change-merged"]) 
```

Is it a bug?

All Buildbot instances uses version: 0.9.15post1

### PR DESCRIPTION
**Pull request by gvalkov**
Set the `files` property for changes emitted by the Gerrit change sources. It's disabled by default as the extra traffic to the Gerrit server may come as an unpleasant surprise to some. The feature can be enabled by setting `get_files` to true.

Fixes #3996.

## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the `master/buildbot/newsfragments` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

**Side-note:**
I last contributed (mostly subversion features and fixes) to buildbot 9 years ago - it's great to see how far the project has come since then! I also can't believe the project is still using some of the [figures](4339e66d11f6f547d9711e14600c86938f1b988d) I made for the documentation back then :)

### CODE REVIEW

File: master/buildbot/changes/gerritchangesource.py [hunk 1]
```diff
@@ -74,6 +75,7 @@ class GerritChangeSourceBase(base.ChangeSource):
     def checkConfig(self,
                     gitBaseURL=None,
                     handled_events=("patchset-created", "ref-updated"),
+                    get_files=False,
```
Comment (sa2ajj) [inline master/buildbot/changes/gerritchangesource.py]: it's a good idea to add new parameters *after* the existing ones: this method signature says that it accepts these parameters as keyword ones as well as positional ones...  i'd even suggest to add `*` before that to ensure that in future this parameter is always a keyword one.

File: master/buildbot/changes/gerritchangesource.py [hunk 2]
```diff
@@ -82,33 +84,37 @@ def checkConfig(self,
     def reconfigService(self,
                         gitBaseURL=None,
                         handled_events=("patchset-created", "ref-updated"),
+                        get_files=False,
```
Comment (sa2ajj) [inline master/buildbot/changes/gerritchangesource.py]: ditto

File: master/buildbot/changes/gerritchangesource.py [hunk 3]
```diff
@@ -82,33 +84,37 @@ def checkConfig(self,
...
                         debug=False):
         self.gitBaseURL = gitBaseURL
         self.handled_events = list(handled_events)
+        self.get_files = get_files
```
Comment (sa2ajj) [inline master/buildbot/changes/gerritchangesource.py]: while we have plenty of "let's have it public data members", i do not think it's a good idea to introduce new ones: please use `self._get_files`

File: master/buildbot/changes/gerritchangesource.py [hunk 4]
```diff
@@ -82,33 +84,37 @@ def checkConfig(self,
...
         self.debug = debug
 
+    @defer.inlineCallbacks
```
Comment (sa2ajj) [inline master/buildbot/changes/gerritchangesource.py]: this change (along with the relevant bits) should belong to a separate commit with explanation what it achieves

File: master/buildbot/changes/gerritchangesource.py [hunk 5]
```diff
@@ -82,33 +84,37 @@ def checkConfig(self,
...
             if self.debug:
                 msg = "no type in event %s"
                 log.msg(msg % line)
-            return defer.succeed(None)
+            return
 
-        return self.eventReceived(event)
+        yield self.eventReceived(event)
 
+    @defer.inlineCallbacks
```
Comment (sa2ajj) [inline master/buildbot/changes/gerritchangesource.py]: ditto

File: master/buildbot/changes/gerritchangesource.py [hunk 6]
```diff
@@ -298,17 +298,30 @@ def streamProcessStopped(self):
...
+        remote host'''
+
+        cmd = [
+            "ssh",
+            "%s@%s" % (self.username, self.gerritserver),
+            "-p", str(self.gerritport)
+        ]
+
+        if self.identity_file is not None:
+            cmd += ["-i", self.identity_file]
```
Comment (sa2ajj) [inline master/buildbot/changes/gerritchangesource.py]: you probably want to use `cmd.extend(...)`

File: master/buildbot/changes/gerritchangesource.py [hunk 7]
```diff
@@ -298,17 +312,49 @@ def streamProcessStopped(self):
...
+
+        if self.debug:
+            log.msg("querying gerrit for changed files in change %s/%s: %s" %
+                    (change, patchset, cmd))
+
+        out = yield utils.getProcessOutput(cmd[0], cmd[1:], env=None)
+        out = out.splitlines()[0]
+        res = json.loads(bytes2unicode(out))
+
+        if res.get("rowCount") == 0:
```
Comment (sa2ajj) [inline master/buildbot/changes/gerritchangesource.py]:
```
if not res.get(...):
```

unless, of course, you want to explicitly catch the case when `rowCount` is present and is equal to 0.

File: master/buildbot/changes/gerritchangesource.py [hunk 8]
```diff
@@ -387,6 +433,14 @@ def eventReceived(self, event):
             yield self.master.db.state.setState(self._oid, 'last_event_ts', event['eventCreatedOn'])
         return res
 
+    @defer.inlineCallbacks
+    def getFiles(self, change, patchset):
+        res = yield self._http.get("/changes/%s/revisions/%s/files/" % (change, patchset))
+        res = yield res.content()
+
+        files = json.loads(bytes2unicode(res.splitlines()[1]))
```
Comment (sa2ajj) [inline master/buildbot/changes/gerritchangesource.py]: do you need this variable here?  why not just `return list(...)`?

File: master/buildbot/changes/gerritchangesource.py [hunk 9]
```diff
@@ -298,17 +312,49 @@ def streamProcessStopped(self):
             self.startStreamProcess()
             self.streamProcessTimeout = self.STREAM_BACKOFF_MIN
 
+    def buildGerritCommand(self, *gerrit_args):
```
Comment (sa2ajj) [inline master/buildbot/changes/gerritchangesource.py]: is this a public method?  if not. use `_buildGerritCommand`

File: master/buildbot/changes/gerritchangesource.py [hunk 10]
```diff
@@ -298,17 +312,49 @@ def streamProcessStopped(self):
...
+    def getFiles(self, change, patchset):
+        cmd = self.buildGerritCommand("query", str(change), "--format", "JSON",
+                                      "--files", "--patch-sets")
+
+        if self.debug:
+            log.msg("querying gerrit for changed files in change %s/%s: %s" %
+                    (change, patchset, cmd))
+
+        out = yield utils.getProcessOutput(cmd[0], cmd[1:], env=None)
+        out = out.splitlines()[0]
```
Comment (sa2ajj) [inline master/buildbot/changes/gerritchangesource.py]: you probably could combine this and the next line

File: master/buildbot/changes/gerritchangesource.py (hunk 1 above)
Comment (gvalkov) [reply master/buildbot/changes/gerritchangesource.py]: Agreed - will fix.

File: master/buildbot/changes/gerritchangesource.py (hunk 3 above)
Comment (gvalkov) [reply master/buildbot/changes/gerritchangesource.py]: Roger that. I was trying to be consistent with the surrounding code (which is a bit of a challenge given that the buildbot sources feature every possible formatting and naming convention out there :)

File: master/buildbot/changes/gerritchangesource.py (hunk 4 above)
Comment (gvalkov) [reply master/buildbot/changes/gerritchangesource.py]: The commits as they are are atomic - I wouldn't be able to split them any further. It's all about making `lineReceived()` preemptable so that `getFiles()` can be called asynchronously. The other option would be to run the query or send the GET request in a way that would block the reactor. I'd be happy to learn if this can be done in another way.

File: master/buildbot/changes/gerritchangesource.py (hunk 6 above)
Comment (gvalkov) [reply master/buildbot/changes/gerritchangesource.py]: Done, though they're equivalent.

File: master/buildbot/changes/gerritchangesource.py (hunk 7 above)
Comment (gvalkov) [reply master/buildbot/changes/gerritchangesource.py]: That is unlikely. If the change exists `gerrit query` will always return two lines of json - one with the result and one with `{"type": "stats", "rowCount": 1, ...}`. If it didn'tt exist then the response would be a single `{"type": "stats", "rowCount": 0}`. I think it's sufficient.

File: master/buildbot/changes/gerritchangesource.py (hunk 8 above)
Comment (gvalkov) [reply master/buildbot/changes/gerritchangesource.py]: Personal style I guess. I often add breakpoints between lines in which case it helps to have a variable ready for inspection. There's a bit too much happening in a line like `list(json.loads(bytes2unicode(res.splitlines()[1])))` - I'll break it down further.

File: master/buildbot/changes/gerritchangesource.py (hunk 6 above)
Comment (sa2ajj) [reply master/buildbot/changes/gerritchangesource.py]: they are indeed, though the new code is consistent with what follows

File: master/buildbot/changes/gerritchangesource.py (hunk 4 above)
Comment (sa2ajj) [reply master/buildbot/changes/gerritchangesource.py]: i am not sure i understand, could you please elaborate?

File: master/buildbot/changes/gerritchangesource.py (hunk 7 above)
Comment (sa2ajj) [reply master/buildbot/changes/gerritchangesource.py]: i'm sorry what is unlikely?

File: master/buildbot/changes/gerritchangesource.py (hunk 4 above)
Comment (gvalkov) [reply master/buildbot/changes/gerritchangesource.py]: Before this change, once a line from `gerrit stream-events` was received by `LocalPP.outReceived` it was passed to `GerritChangeSourceBase.lineReceived` and execution proceeded sequentially/synchronously all the way to `addChangeFromEvent`. 

However, we now want to perform a costly blocking operation (i.e `getFiles`) in the middle of all this. By making `addChangeFromEvent` an async function, we allow the buildbot master to do something else useful while `getFiles()` is waiting for its result to come back (I've seen cold queries for changes with many patchsets take up to half a second to return). For this to work, we have to make the whole call chain up to `lineReceived` async for this to work - i.e. `addChangeFromEvent, eventReceived, lineReceived` must all return deferreds.

File: master/buildbot/changes/gerritchangesource.py (hunk 7 above)
Comment (gvalkov) [reply master/buildbot/changes/gerritchangesource.py]: It's just that if the first line of output is `{"type": "stats", "rowCount": 0}` then no results were found, i.e. the change doesn't exist, which is unlikely given that we're reacting to an event about an existing change.

File: master/buildbot/changes/gerritchangesource.py (hunk 9 above)
Comment (gvalkov) [reply master/buildbot/changes/gerritchangesource.py]: Done

File: master/buildbot/changes/gerritchangesource.py (hunk 4 above)
Comment (tardyp) [reply master/buildbot/changes/gerritchangesource.py]: It is not really needed to turn this code into inlineCallbacks. The code is already asynchronous as it return defer.succeed, and returns the  value of self.eventReceived (which could be a deferred, generated by an inlineCallback generator)

File: master/buildbot/changes/gerritchangesource.py [hunk 11]
```diff
@@ -387,6 +433,14 @@ def eventReceived(self, event):
             yield self.master.db.state.setState(self._oid, 'last_event_ts', event['eventCreatedOn'])
         return res
 
+    @defer.inlineCallbacks
+    def getFiles(self, change, patchset):
+        res = yield self._http.get("/changes/%s/revisions/%s/files/" % (change, patchset))
+        res = yield res.content()
+
+        res = bytes2unicode(res.splitlines()[1])  # the first line of every response is `)]}'`
```
Comment (tardyp) [inline master/buildbot/changes/gerritchangesource.py]: you can just call .decode() now that we don't have to deal with python2.6

Comment (gvalkov) [reply master/buildbot/changes/gerritchangesource.py]: Good news. I would have removed it in the other places where it's called, but several of the tests  depend on the "passthrough" behaviour of `bytes2unicode`. Can fix in a follow up PR.

File: master/buildbot/changes/gerritchangesource.py (hunk 4 above)
Comment (gvalkov) [reply master/buildbot/changes/gerritchangesource.py]: You're right. I initially got into a big mess that I could resolve only by using `inlineCallbacks` on everything down the call chain. It should be simpler now (i.e. `inlineCallback` only on `addChangeFromEvent` since that calls `getFiles`).

**Comment by gvalkov** [comment]
Hi @tardyp . I've simplified the change (no unneeded inlineCallbacks-ification). More over I've been running the change cherry-picked on top of the 2.3.1 release in our production buildbot instance since I opened the PR. Is there anything else I can do to help with merging this change?
